config/load.go

```diff
@@ -0,0 +1,89 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+)
+
+// Getter is an interface for getting
+// config file text
+type Getter interface {
+	GetConfig() ([]byte, error)
+}
+
+// Unmarshaller is an interface for
+// unmarshalling config text into a destination
+type Unmarshaller interface {
+	Unmarshal(data []byte, v any) error
+}
+
+// Loader is a struct used to get
+// and unmarshal config data
+type Loader struct {
+	Getters       []Getter
+	Unmarshallers []Unmarshaller
+}
+
+var (
+	ErrFailedToGetConfig       = fmt.Errorf("failed to get config")
+	ErrFailedToUnmarshalConfig = fmt.Errorf("failed to unmarshal config")
+)
+
+// Load gets the config from the first successful getter.Get()
+// then unmarshals it using the first successful
+// unmarshaller.Unmarshal
+func (o *Loader) Load(v any) error {
+	var (
+		cfgData []byte
+		err     error
+	)
+
+	for _, getter := range o.Getters {
+		cfgData, err = getter.GetConfig()
+		if err == nil {
+			break
+		}
+	}
+
+	if err != nil {
+		return errors.Join(ErrFailedToGetConfig, err)
+	}
+
+	for _, unmarshaller := range o.Unmarshallers {
+		err = unmarshaller.Unmarshal(cfgData, v)
+		if err == nil {
+			break
+		}
+	}
+
+	if err != nil {
+		return errors.Join(ErrFailedToUnmarshalConfig, err)
+	}
+
+	return nil
+}
+
+// NewToolsConfigLoaderFile creates a new ConfigLoader with config
+// getters and unmarshalers matching what the Aerospike Tools config files support.
+func NewToolsConfigLoaderFile(configPath string) *Loader {
```

Comment:
I think the `New*` function should be the first method under the struct definition. That seems to be a common pattern among existing files and other projects

Reply:
In the past I've done the opposite, I usually like to keep types and their methods close together. I can't find any style guide that comments on this and I see both ways used in the golang source code.

I'm ok with using the style you suggest I just think we should standardize on something. I think we need an internal go style guide for stuff like this. https://aerospike.atlassian.net/wiki/spaces/TOOLS/pages/3227189281/Golang+Style+Guide

Reply:
I think when in doubt stick to the style of the code that already exists. There are a few examples of this already